File: Aditya Practicals/MCM.cpp
#include <bits/stdc++.h>
using namespace std;

void print_tables(int table[][100], int n) {
    for(int i=1; i<=n; i++) {
        for(int j=1; j<=n; j++) {
            cout << setw(10) << table[i][j] << " ";
        }
        cout << endl;
    }
}

int matrix_chain_order(int p[], int n) {
    int m[100][100], s[100][100];
    for(int i=1; i<n; i++) m[i][i] = 0;
    for(int L=2; L<n; L++) {
        for(int i=1; i<n-L+1; i++) {
            int j = i+L-1;
            m[i][j] = INT_MAX;
            for(int k=i; k<=j-1; k++) {
                int q = m[i][k] + m[k+1][j] + p[i-1]*p[k]*p[j];
                if(q < m[i][j]) {
                    m[i][j] = q;
                    s[i][j] = k;
                }
            }
        }
    }
    //cout<<INT_MAX<<endl;
    cout << "m table:\n";
    print_tables(m, n-1);
    cout << "\ns table:\n";
    print_tables(s, n-1);
    return m[1][n-1];
}

int main() {
    int n;
    cout << "Enter the number of matrices: ";
    cin >> n;
    int p[n+1];
    cout << "Enter the dimensions of the matrices:\n";
    for(int i=0; i<=n; i++) {
        cin >> p[i];
    }
    cout << "Minimum cost for matrix multiplication: " << matrix_chain_order(p, n+1) << endl;
    return 0;
}

File: Aditya Practicals/mcm.py
import sys

def print_tables(table, n):
    for i in range(1, n+1):
        for j in range(1, n+1):
            print("{:10d}".format(table[i][j]), end=" ")
        print()

def print_optimal_parentheses(s, i, j):
    if i == j:
        print("A{}".format(i), end="")
    else:
        print("(", end="")
        print_optimal_parentheses(s, i, s[i][j])
        print_optimal_parentheses(s, s[i][j]+1, j)
        print(")", end="")

def matrix_chain_order(p, n):
    m = [[0 for _ in range(n)] for _ in range(n)]
    s = [[0 for _ in range(n)] for _ in range(n)]
    for i in range(1, n):
        m[i][i] = 0
    for L in range(2, n):
        for i in range(1, n-L+1):
            j = i+L-1
            m[i][j] = sys.maxsize
            for k in range(i, j):
                q = m[i][k] + m[k+1][j] + p[i-1]*p[k]*p[j]
                if q < m[i][j]:
                    m[i][j] = q
                    s[i][j] = k
    print("m table:")
    print_tables(m, n-1)
    print("\ns table:")
    print_tables(s, n-1)
    print("\nOptimal Parentheses: ", end="")
    print_optimal_parentheses(s, 1, n-1)
    print()
    return m[1][n-1]

n = int(input("Enter the number of matrices: "))
p = list(map(int, input("Enter the dimensions of the matrices:\n").split()))
print("Minimum cost for matrix multiplication:", matrix_chain_order(p, n+1))
